FEAT: more functionality for highlighting MLIR/xDSL Code

webmiche opened the pull request:
This PR adds support for syntax highlighting more IR styles, in particular xDSL style. As I did not follow MLIR's guidelines very closely but rather implemented syntax highlighting for the code i need in my thesis, this might support some ill-formatted IRs. However, I still wanted to make this available for more people.

I feel that we should probably refactor this majorly in order to make all the regex readable and understand what they are supposed to match on.

tobiasgrosser commented:
Nice. Thank you for adding this. We might need refactoring and maybe even some test cases here. Now, maybe it is better to leave this for a while as it is. I hope we will add syntax support to IRDL in a couple of months. I would then expect that we can automatically derive pygment definitions for different dialects. For now, I feel we can merge what you have.

webmiche commented:
I moved some rules out in order to give them names. I feel that makes working with the rules and extending them far easier. What do you think? Obviously, this approach will not work for things like the `MlirLexerOnlyOps` defined below...

math-fehr reviewed (commented):
That seems good to me!
Do you think you could add an example you have from your thesis? Just a simple program so we are sure we won't break it if we modify something else?

webmiche commented:
I added an example. Notice that I have changed the syntax slightly at some points in order to not be as verbose as the actual print. In particular, I have removed the `""` from the attribute names and I sometimes use `...` to not have to write out all types.

I feel like printing in this way looks better for a paper, so I would argue to keep it, but I can adjust the lexer if you want to have syntax highlighting for the standard IR prints.

math-fehr commented:
Looks good, thanks!
